alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -23,25 +54,143 @@ class BlackboardImpl
+        }
     }
     template <typename T>
     T& get(const std::string& key)
     {
-        return std::any_cast<T&>(vals.at(key));
+        try {
+            return std::get<T>(vals.at(key));
+        } catch (const std::bad_variant_access& e) {
+            throw BlackboardTypeMismatch(getNameFromType<T>(), getBlackboardValueType(key));
+        }
     }
-    std::any& get(const std::string& key) { return vals.at(key); }
-    const std::any& get(const std::string& key) const { return vals.at(key); }
+    BlackboardValue& get(const std::string& key) { return vals.at(key); }
+    const BlackboardValue& get(const std::string& key) const { return vals.at(key); }
 
-    void set(const std::string& key, const std::any& value) { vals.at(key) = value; }
+    template <class T>
+    void set(const std::string& key, const T& value)
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        if (entry[Strings::key].as<std::string>() != key) {
+            // value is not predefined in pml file
+            vals.at(key) = value;
+        } else {
+            // value is predefined in pml file
+            std::string inputType = getNameFromType<T>();
+            std::string pmlType = getBlackboardValueType(key);
+            if (pmlType == inputType) {
+                vals.at(key) = value;
+            } else {
+                // types dont match, throw exception
+                throw BlackboardTypeMismatch(inputType, pmlType);
+            }
+        }
+    }
 
     bool hasValue(const std::string& key) const { return vals.count(key); }
     void removeValue(const std::string& key) { vals.erase(key); }
 
+    void initDefaultValues();
+
     void clear() { vals.clear(); }
     bool empty() const { return vals.empty(); }
     size_t size() const { return vals.size(); }
-    std::unordered_map<std::string, std::any> vals;
+    std::unordered_map<std::string, BlackboardValue> vals;
+    YAML::Node node;
+
+private:
+    friend BlackboardUtil;
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
+            if (entry[Strings::key].as<std::string>() == key) {
+                return entry;
+            }
+        }
+        return YAML::Node();
+    }
+
+    std::string getBlackboardValueType(const std::string& key) const
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        return entry[Strings::stateType].as<std::string>();
+    }
+
+    template <class T>
+    std::string getNameFromType() const;
+};
+
+struct Converter
+{
+    static constexpr const char* typeNames[] = {"bool", "int64", "double", "std::string", "std::any"};
+
+    // YAML doesnt support std::any, needs a separate list of types for iteration
+    static constexpr const char* typeNamesYAML[] = {"bool", "int64", "double", "std::string"};
+    using TypesYAML = std::variant<bool, int64_t, double, std::string>;
```

Comment:
Can't we get rid of this as well? The type of the default value should be one of the strings in `typeNames[]` (apart from `std::any`). If it is not, we simply reject setting the value (for now).